tools/update_lint_rules/lib/src/mappers/lint_code_dto_mapper.dart

```diff
@@ -0,0 +1,81 @@
+import 'package:collection/collection.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:update_lint_rules/src/models/lint_code_dto.dart';
+import 'package:update_lint_rules/src/models/lint_rule.dart';
+
+class LintCodeDtoMapper {
+  const LintCodeDtoMapper._();
+
+  /// Builds a [Rule] instance from provided parameters.
+  ///
+  /// If any of the required parameters categories, details, or state
+  /// is null. The [FormatException] message includes which fields are missing.
+  @visibleForTesting
+  static Rule buildRule({
+    required String name,
+    List<String>? categories,
+    String? details,
+    Map<String, dynamic>? state,
+  }) {
+    if (categories == null || details == null || state == null) {
+      throw FormatException(
+        'The required field for the $name is null: ${[if (categories == null) 'categories', if (details == null) 'details', if (state == null) 'state'].join(', ')}',
+      );
+    }
+
+    return Rule(
+      name: name,
+      categories: categories,
+      details: details,
+      state: state.map(
+        (key, value) =>
+            MapEntry(RuleState.values.byName(key), Since.fromJson(value)),
+      ),
+    );
+  }
+
+  /// Convert a list of [LintCodeDto] to a list of [Rule]
+  static List<Rule> toRules(Iterable<LintCodeDto> dtos) {
+    // Group DTOs by sharedName
+    final groupedLintCodeDtosBySharedName = dtos
+        .where((dto) => dto.sharedName != null)
+        // If `dto.sharedName` is not null, `dto.name` is the same as `dto.sharedName`.
+        // So, group by `dto.name`.
+        .groupListsBy((dto) => dto.name);
+
+    // Convert DTOs with sharedName to Rules
+    final rulesWithSharedName = groupedLintCodeDtosBySharedName.entries.map((
+      e,
+    ) {
+      final dtos = e.value;
+
+      final categories = dtos.map((e) => e.categories).nonNulls.firstOrNull;
+      final deprecatedDetails =
+          dtos.map((e) => e.deprecatedDetails).nonNulls.firstOrNull;
+      final state = dtos.map((e) => e.state).nonNulls.firstOrNull;
+
+      return LintCodeDtoMapper.buildRule(
+        name: e.key,
+        categories: categories,
+        details: deprecatedDetails,
+        state: state,
+      );
+    });
+
+    // Convert DTOs without sharedName to Rules
+    final rulesWithoutSharedName = dtos
+        .where((dto) => dto.sharedName == null)
+        .map(
+          (dto) => LintCodeDtoMapper.buildRule(
+            name: dto.name,
+            categories: dto.categories ?? const [],
```

Comment:
Since data such as [always_require_non_null_named_parameters](https://github.com/dart-lang/sdk/blob/ae6da8b926f208bf87d2e11375be5c611c27ee1b/pkg/linter/messages.yaml#L222-L250)'s for which `category` is null already exists, an empty character is assigned when `category` is null.